### PR TITLE
run-make tests initialized

### LINKED ISSUE #263: Run other types of tests like run-make
`run-make` is a testsuite [from the Rust repo](https://github.com/rust-lang/rust/tree/master/tests/run-make).

There are other [tests from the Rust testsuite](https://github.com/rust-lang/rust/tree/master/tests) that we could run as well.

It probably needs to be done in a similar way to [how we do the UI tests](https://github.com/rust-lang/rustc_codegen_gcc/blob/06d07f816a9e95de360c95cabca907bd6e666905/build_system/src/test.rs#L1080-L1094).

### PR DESCRIPTION
**Pull request by lordshashank**
- resolves https://github.com/rust-lang/rustc_codegen_gcc/issues/263
- Added run-make tests
- made failing-run-make-tests.txt for failing tests.

### CODE REVIEW

File: tests/failing-run-make-tests.txt [hunk 1]
```diff
@@ -30,4 +30,13 @@ tests/run-make/rustdoc-themes
...
+tests/run-make/rustdoc-with-short-out-dir-option
+tests/run-make/return-non-c-like-enum
+tests/run-make/lto-smoke-c
+tests/run-make/issue-68794-textrel-on-minimal-lib
+tests/run-make/issue-36710
+tests/run-make/glibc-staticlib-args
+tests/run-make/foreign-exceptions
+tests/run-make/foreign-double-unwind
+tests/run-make/c-link-to-rust-staticlib
+tests/run-make/arguments-non-c-like-enum
```
Comment (antoyo) [inline tests/failing-run-make-tests.txt]: It seems you might only need to install the package to be able to link `lstdc++` to have these tests pass in the CI.

Comment (lordshashank) [reply tests/failing-run-make-tests.txt]: @antoyo sorry, I don't understand, what package are you referring to? 
And would this resolve only these 9 or all failing tests?

Comment (antoyo) [reply tests/failing-run-make-tests.txt]: The package is named either `libstdc++-12-dev` or `libstdc++-11-dev`.
I only saw the failures for those 9 tests, so probably only these one (perhaps not all of these 9 tests).

**Comment by lordshashank** [comment]
@antoyo added libstdc++-12-dev (if this is how you meant), also removed all failing tests so that I can get all the tests failing in ci, independent of my local system, hope thats not an issue.

**Comment by antoyo** [comment]
It's good. Thanks!

**Comment by antoyo** [comment]
Weird. We still get the error "/usr/bin/ld: cannot find -lstdc++: No such file or directory". Did you ever get that locally?

**Comment by lordshashank** [comment]
@antoyo I am thinking to deprecate version to 11, and putting this in ci to debug
```
    - name: Install packages
      # `llvm-14-tools` is needed to install the `FileCheck` binary which is used for asm tests.
      run: |
          sudo apt-get install ninja-build ripgrep llvm-14-tools libstdc++-11-dev
          locate libstdc++
          echo $LD_LIBRARY_PATH
 ```
 should I do that (not sure we should do echo in ci runs)?
 ALso do review my code, in case there's some mistake, there.

**Comment by antoyo** [comment]
You want to try version 11 to see if this will fix the tests? If so, yes, this is a good idea.

I wanted to review the PR when the CI pass.
The PR mostly looks good: the changes I would want to see are less code duplication because we'll need to do the same for other test suites in Rust like codegen-units,run-pass-valgrind, and incremental.
So, functions like `test_rustc_inner_run_make` and the callbacks sent to this function should get refactored to remove code duplication with `test_rustc_inner_ui`.

Another thing I'm still thinking about is whether we split those new tests like `run-make` since there seems to be much less of them than UI tests.

**Comment by lordshashank** [comment]
@antoyo I tried to reduce duplication by modularising some code (thought a lot, let me know if you have better way). Also I tried to remove the ui tests [here ](https://github.com/lordshashank/rustc_codegen_gcc/blob/5918a517c0d614582b6cf4ad2590f53dac18b459/build_system/src/test.rs#L921C1-L930C1) to failing-ui-tests, but ci failed so reseted them, but CI still fails. These are running in my device successfully, not getting why are they failing here, could you please help.

**Comment by antoyo** [comment]
It looks like some test files are being removed twice, causing this error:
```
Command failed to run: Failed to remove `build/rust/tests/ui/impl-trait/equality-in-canonical-query.rstests/ui/asm/x86_64/issue-96797.rs`: Os { code: 2, kind: NotFound, message: "No such file or directory" }
```

Since we're removing files, I guess we might need to do a `git checkout` somewhere between the two runs to restore the files.

**Comment by lordshashank** [comment]
@antoyo I am not able to get the issue here as both the runs are failing here, moreover I am not able to replicate this, any views.

**Comment by antoyo** [comment]
Would you be OK if I push commits to this PR to attempt to fix the issue?

**Comment by lordshashank** [comment]
ya sure 😓, there have ben too many commits, should I squash them first?

**Comment by antoyo** [comment]
Yes please.

**Comment by lordshashank** [comment]
Ah! wait I guess I got it

**Comment by lordshashank** [comment]
@antoyo , you can do changes, actually its showing the change in same file in failing-ui-tests.txt , but its irrelevant i guess cause it shows same line added and removed, some ci issue i guess, look forward to what solution is.

**Comment by lordshashank** [comment]
could shifting this test to test.rs for ui-tests from failing-ui-tests.txt resolve the issue?, cause its this test only which is giving issue on all failing ci runs (I rechecked them)

File: tests/failing-ui-tests.txt [hunk 2]
```diff
@@ -75,4 +75,4 @@ tests/ui/async-await/in-trait/dont-project-to-specializable-projection.rs
 tests/ui/consts/try-operator.rs
 tests/ui/coroutine/unwind-abort-mix.rs
 tests/ui/type-alias-impl-trait/rpit_tait_equality_in_canonical_query.rs
-tests/ui/impl-trait/equality-in-canonical-query.rs
+tests/ui/impl-trait/equality-in-canonical-query.rs
```
Comment (antoyo) [inline tests/failing-ui-tests.txt]: The error is caused by this change: it seems your text editor removed the newline at the end, so please revert this change.

Comment (lordshashank) [reply tests/failing-ui-tests.txt]: This one space wasted one day 😭😭

**Comment by lordshashank** [comment]
@antoyo, so checks are successful finally, let me know of any more improvements, 
One I wanted your view on was I removed the tests splitting for run-make tests, but now they are getting ran on both runs unnecessarily, should I reinstate splitting or put a check to run run-make test only when current_parts is 0? (I think later one would be good)

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -47,7 +47,8 @@ jobs:
 
     - name: Install packages
       # `llvm-14-tools` is needed to install the `FileCheck` binary which is used for asm tests.
-      run: sudo apt-get install ninja-build ripgrep llvm-14-tools
+      run: |
+          sudo apt-get install ninja-build ripgrep llvm-14-tools libstdc++-11-dev
```
Comment (antoyo) [inline .github/workflows/ci.yml]: Are you sure this change is needed?
I would expect it to be required in some other workflows if it was the case, but all the CI passed, so maybe it's not actually needed.

Comment (lordshashank) [reply .github/workflows/ci.yml]: not needed, I added tests failing due to this on txt file, would remove this. We could create a different issue for that linker error as well imo to add libstdc++.

File: build_system/src/test.rs [hunk 4]
```diff
@@ -856,134 +856,135 @@ fn should_remove_test(file_path: &Path) -> Result<bool, String> {
...
-            remove_file(&rust_path.join(path))?;
+        remove_file(&rust_path.join("tests/ui/consts/const_cmp_type_id.rs"))?;
+        remove_file(&rust_path.join("tests/ui/consts/issue-73976-monomorphic.rs"))?;
+        // this test is oom-killed in the CI.
+        remove_file(&rust_path.join("tests/ui/consts/issue-miri-1910.rs"))?;
+        // Tests generating errors.
+        remove_file(&rust_path.join("tests/ui/consts/issue-94675.rs"))?;
+        remove_file(&rust_path.join("tests/ui/mir/mir_heavy_promoted.rs"))?;
+        remove_file(&rust_path.join("tests/ui/rfcs/rfc-2632-const-trait-impl/const-drop-fail.rs"))?;
+        remove_file(&rust_path.join("tests/ui/rfcs/rfc-2632-const-trait-impl/const-drop.rs"))?;
```
Comment (lordshashank) [inline build_system/src/test.rs]: Can these be added to failing-ui-tests file. It would be better, not sure why they are here.

Comment (antoyo) [reply build_system/src/test.rs]: Yes, I think we can move them. Or even better, it seems some of these tests have the annotation `// @known-bug` in them, so if this annotation is indeed indicating a known failure, we can add it [here](https://github.com/rust-lang/rustc_codegen_gcc/blob/master/build_system/src/test.rs#L914).

File: build_system/src/test.rs [hunk 5]
```diff
@@ -856,134 +856,135 @@ fn should_remove_test(file_path: &Path) -> Result<bool, String> {
...
-                    format!("Failed to remove folder `{}`: {:?}", dir.display(), error)
-                })?;
-            }
-            Ok(())
-        },
-        |_| Ok(()),
-    )?;
-
-    // These two functions are used to remove files that are known to not be working currently
-    // with the GCC backend to reduce noise.
```
Comment (antoyo) [inline build_system/src/test.rs:894]: Please keep this comment.

File: build_system/src/test.rs [hunk 6]
```diff
@@ -856,134 +856,135 @@ fn should_remove_test(file_path: &Path) -> Result<bool, String> {
...
+                        format!("Failed to remove folder `{}`: {:?}", dir.display(), error)
+                    })?;
+                }
+                Ok(())
+            },
+            |_| Ok(()),
+        )?;
 
-    if !prepare_files_callback(&rust_path)? {
-        // FIXME: create a function "display_if_not_quiet" or something along the line.
```
Comment (antoyo) [inline build_system/src/test.rs:927]: Please also keep all the `FIXME` comments.

File: build_system/src/test.rs [hunk 7]
```diff
@@ -856,134 +856,135 @@ fn should_remove_test(file_path: &Path) -> Result<bool, String> {
...
-        .map_err(|error| format!("Failed to retrieve output of find command: {:?}", error))?;
-        let mut files = out
-            .split('\n')
-            .map(|line| line.trim())
-            .filter(|line| !line.is_empty())
-            .collect::<Vec<_>>();
-        // To ensure it'll be always the same sub files, we sort the content.
-        files.sort();
-        // We increment the number of tests by one because if this is an odd number, we would skip
-        // one test.
```
Comment (antoyo) [inline build_system/src/test.rs:962]: Please add back all the comments that were removed as it helps understand the code.

File: build_system/src/test.rs [hunk 8]
```diff
@@ -856,134 +856,135 @@ fn should_remove_test(file_path: &Path) -> Result<bool, String> {
...
 
     let extra =
         if args.is_using_gcc_master_branch() { "" } else { " -Csymbol-mangling-version=v0" };
 
     let rustc_args = format!(
-        "{test_flags} -Zcodegen-backend={backend} --sysroot {sysroot}{extra}",
-        test_flags = env.get("TEST_FLAGS").unwrap_or(&String::new()),
-        backend = args.config_info.cg_backend_path,
-        sysroot = args.config_info.sysroot_path,
-        extra = extra,
```
Comment (antoyo) [inline build_system/src/test.rs]: Please keep those arguments named as it helps readability.

File: build_system/src/test.rs [hunk 9]
```diff
@@ -1003,54 +1004,153 @@ where
...
+        args,
+        prepare_files_callback_failing("tests/failing-run-make-tests.txt", "run-make"),
+        "run-make",
+    )?;
+    test_rustc_inner(
+        env,
+        args,
+        prepare_files_callback_failing("tests/failing-ui-tests.txt", "ui"),
+        "ui",
+    )
```
Comment (antoyo) [inline build_system/src/test.rs]: It seems the failing [UI tests are not ran in the CI](https://github.com/antoyo/rustc_codegen_gcc/actions/runs/8409674523/job/23027225505?pr=7#step:15:2919).

Please check if this is the case and fix the issue if there's one.

File: build_system/src/test.rs [hunk 10]
```diff
@@ -1003,54 +1004,153 @@ where
...
+}
+
+fn prepare_files_callback_failing<'a>(
+    file_path: &'a str,
+    test_type: &'a str,
+) -> impl Fn(&Path) -> Result<bool, String> + 'a {
+    move |rust_path| {
+        let files = std::fs::read_to_string(file_path).unwrap_or_default();
+        let first_file_name = files.lines().next().unwrap_or("");
```
Comment (antoyo) [inline build_system/src/test.rs:1050]: It wasn't obvious to me at first that this will apply to all lines in the file.

Could you please explain why you decided to exclude only whole directories for the `run-make` tests?
Is it because all the tests in all these directories failed?

File: build_system/src/test.rs [hunk 11]
```diff
@@ -1003,54 +1004,153 @@ where
...
+                    files.split('\n').map(|line| line.trim()).filter(|line| !line.is_empty())
+                {
+                    run_command(&[&"git", &"checkout", &"--", &file], Some(&rust_path))?;
+                }
+            } else {
+                println!(
+                    "Failed to read `{}`, not putting back failing `{}` tests",
+                    file_path, test_type
+                );
+            }
```
Comment (antoyo) [inline build_system/src/test.rs]: It seems like this code is duplicated in the `if` body above.
Please put it after the `if/else` to avoid this duplication.

**Review by antoyo — COMMENTED** [review]
Very nice work!

File: build_system/src/test.rs (hunk 10 above)
Comment (lordshashank) [reply build_system/src/test.rs:1050]: Ya, I ran all the tests and the logs of failing tests indicated to directories only (which were files in ui tests), I checked few of them, contains single tests on makefile, thus I included the folders. let me know if I missed something error.

Comment (antoyo) [reply build_system/src/test.rs:1050]: In this case, please add a comment to say that we assume that all the lines are directory if the first line ends with `/`.

File: build_system/src/test.rs (hunk 9 above)
Comment (lordshashank) [reply build_system/src/test.rs]: @antoyo I got around 170 tests on them, should I remove the pattern and add those to failing test file? Not sure I get what you meant here. CI runs only successful tests imo

**Comment by lordshashank** [comment]
> @antoyo, so checks are successful finally, let me know of any more improvements, One I wanted your view on was I removed the tests splitting for run-make tests, but now they are getting ran on both runs unnecessarily, should I reinstate splitting or put a check to run run-make test only when current_parts is 0? (I think later one would be good)

@antoyo should I do this?

File: build_system/src/test.rs (hunk 9 above)
Comment (antoyo) [reply build_system/src/test.rs]: What I mean is that this should run the failing `make-run` tests in addition to the failing `ui` tests, but it seems the `ui` tests are not ran in the CI.

Perhaps the function does an early return because of `?` so the `ui` tests are not ran.

**Comment by antoyo** [comment]
> > @antoyo, so checks are successful finally, let me know of any more improvements, One I wanted your view on was I removed the tests splitting for run-make tests, but now they are getting ran on both runs unnecessarily, should I reinstate splitting or put a check to run run-make test only when current_parts is 0? (I think later one would be good)
> 
> @antoyo should I do this?

I'll think about this. The reason why the `ui` tests are splitted is to make the CI faster, but this doesn't seem necessary for the `make-run` tests.

You might not know, but there's a [summary of failing tests here](https://github.com/antoyo/rustc_codegen_gcc/actions/runs/8410624921?pr=7#summary-23029297572) which shows the number of failing tests for the Failures CI task.
With your PR, the results shown there only show the failures for the `make-run` tests: perhaps this is [because the `ui` failing tests do not seem to be ran](https://github.com/antoyo/rustc_codegen_gcc/pull/7#discussion_r1536839111).

I'd like to see how this will render in the summary when those `ui` failing tests are ran in the CI: if it looks good, perhaps we can just keep the `make-run` tests the way it is (ran twice).

**Comment by lordshashank** [comment]
all tests successful LFG 🚀. @antoyo let me know if some more changes are needed.

File: build_system/src/test.rs [hunk 12]
```diff
@@ -840,6 +840,7 @@ fn should_remove_test(file_path: &Path) -> Result<bool, String> {
             "-Cllvm-args",
             "//~",
             "thread",
+            "//@ known-bug"
```
Comment (antoyo) [inline build_system/src/test.rs]: Actually, I was wrong on this one as it seems some of those tests can pass.
Please remove it and add the tests to `failing-ui-tests.txt`.
Sorry.

File: build_system/src/test.rs [hunk 13]
```diff
@@ -865,111 +871,104 @@ where
...
-    if !prepare_files_callback(&rust_path)? {
-        // FIXME: create a function "display_if_not_quiet" or something along the line.
-        println!("Keeping all UI tests");
-    }
+        walk_dir(rust_path.join("tests/ui"), dir_handling, file_handling)?;
 
-    let nb_parts = args.nb_parts.unwrap_or(0);
-    if nb_parts > 0 {
-        let current_part = args.current_part.unwrap();
-        // FIXME: create a function "display_if_not_quiet" or something along the line.
```
Comment (antoyo) [inline build_system/src/test.rs:934]: Please add back this comment.

File: build_system/src/test.rs [hunk 14]
```diff
@@ -1045,6 +1047,7 @@ fn prepare_files_callback_failing<'a>(
     move |rust_path| {
         let files = std::fs::read_to_string(file_path).unwrap_or_default();
         let first_file_name = files.lines().next().unwrap_or("");
+        // If the first line ends with a `/`, we treat it as a directory.
```
Comment (antoyo) [inline build_system/src/test.rs]: Please rephrase to something like this:
```
// If the first line ends with a `/`, we treat all lines in the file as a directory.
```

File: build_system/src/test.rs [hunk 15]
```diff
@@ -1003,54 +1003,144 @@ where
...
-                let path = rust_path.join(file);
-                remove_file(&path)?;
+fn prepare_files_callback_success<'a>(
+    file_path: &'a str,
+    test_type: &'a str,
+) -> impl Fn(&Path) -> Result<bool, String> + 'a {
+    move |rust_path| {
+        let files = std::fs::read_to_string(file_path).unwrap_or_default();
+        let first_file_name = files.lines().next().unwrap_or("");
+        // If the first line ends with a `/`, we treat it as a directory.
```
Comment (antoyo) [inline build_system/src/test.rs]: Same here:
```suggestion
        // If the first line ends with a `/`, we treat all lines in the file as a directory.
```

**Review by antoyo — COMMENTED** [review]
I missed a few things.
After that, we can merge.
Thanks a lot!

File: build_system/src/test.rs (hunk 12 above)
Comment (lordshashank) [reply build_system/src/test.rs]: Ahhh! yes, I missed this as well.

File: build_system/src/test.rs (hunk 14 above)
Comment (lordshashank) [reply build_system/src/test.rs]: yeah better.

File: build_system/src/test.rs [hunk 16]
```diff
@@ -839,8 +839,7 @@ fn should_remove_test(file_path: &Path) -> Result<bool, String> {
             "//@ run-fail",
             "-Cllvm-args",
             "//~",
-            "thread",
-            "//@ known-bug"
+            "thread"
```
Comment (antoyo) [inline build_system/src/test.rs]: Please keep the trailing comma.

**Review by antoyo — COMMENTED** [review]
One more nitpick ;) .

**Comment by lordshashank** [comment]
Really sorry @antoyo , I don't know why that spacing at last got ommitted even though I tried to add it 😭, this should work now.

**Comment by antoyo** [comment]
> Really sorry @antoyo , I don't know why that spacing at last got ommitted even though I tried to add it 😭, this should work now.

No worries.
I know that some text editor removes the last new line in files. Perhaps your text editor has a setting to configure this behavior?

File: build_system/src/test.rs [hunk 17]
```diff
@@ -865,111 +870,105 @@ where
...
-            .map(|line| line.trim())
-            .filter(|line| !line.is_empty())
-            .collect::<Vec<_>>();
-        // To ensure it'll be always the same sub files, we sort the content.
-        files.sort();
-        // We increment the number of tests by one because if this is an odd number, we would skip
-        // one test.
-        let count = files.len() / nb_parts + 1;
-        let start = current_part * count;
-        // We remove the files we don't want to test.
```
Comment (antoyo) [inline build_system/src/test.rs:965]: Please add back this comment.

File: build_system/src/test.rs [hunk 18]
```diff
@@ -1003,54 +1003,144 @@ where
...
         }
+
         Ok(true)
-    })
+    }
 }
 
-fn test_successful_rustc(env: &Env, args: &TestArg) -> Result<(), String> {
-    test_rustc_inner(env, args, |rust_path| {
-        // Removing the failing tests.
```
Comment (antoyo) [inline build_system/src/test.rs:1042]: Please add back this comment.

**Review by antoyo — COMMENTED** [review]
Sorry, I missed those two comments.
After that, we can merge for real :) .

**Comment by lordshashank** [comment]
Done and sorry, I should not have removed the comments, would remember for next time.

**Review by antoyo — APPROVED** [review]

**Comment by lordshashank** [comment]
@antoyo , what other test suites of rust do you look forward to implement, I can try them as I guess they would be fairly easy to implement after this modularization.

**Comment by antoyo** [comment]
I would find the most important to be the standard library (std) tests.
